### PR TITLE
add Shop_UnregisterItemId native

### PR DESCRIPTION
**Pull request by IL0co**
Adding a native to unload an item by its ItemId

### CODE REVIEW

**Review by R1KO — COMMENTED** [review]
@TiBarification @CrazyHackGUT

Название `Shop_UnregisterItemId` норм или лучше  `Shop_UnregisterItem` ?

**Comment by CrazyHackGUT** [comment]
Второе, думаю.

**Comment by TiBarification** [comment]
Второе

File: addons/sourcemod/scripting/shop/item_manager.sp [hunk 1]
```diff
@@ -45,6 +45,8 @@ void ItemManager_CreateNatives()
 	CreateNative("Shop_SetCallbacks", ItemManager_SetCallbacks);
 	CreateNative("Shop_SetHide", ItemManager_SetHide);
 	CreateNative("Shop_EndItem", ItemManager_EndItem);
+
+	CreateNative("Shop_UnregisterItemId", ItemManager_UnregisterItem);
```
Comment (TiBarification) [inline addons/sourcemod/scripting/shop/item_manager.sp:49]: Change it to "Shop_UnregisterItem", so it not be so weird.

File: addons/sourcemod/scripting/include/shop/items.inc [hunk 2]
```diff
@@ -203,6 +203,15 @@ native bool Shop_SetItemCustomInfoString(ItemId item_id, const char[] info, cons
...
 native bool Shop_KvCopySubKeysItemCustomInfo(ItemId item_id, KeyValues kv);
 
+/**
+ *	Unregisters item, categories and removes them from the shop and players' inventory by ItemId
+ *	
+ *	@param item_id				Item id
+ *	
+ *	@noreturn
+*/
+native void Shop_UnregisterItemId(ItemId item_id);
```
Comment (TiBarification) [inline addons/sourcemod/scripting/include/shop/items.inc:213]: Change it to "Shop_UnregisterItem", so it not be so weird.

**Review by TiBarification — CHANGES_REQUESTED** [review]

File: addons/sourcemod/scripting/shop/item_manager.sp [hunk 3]
```diff
@@ -576,6 +578,27 @@ public int ItemManager_EndItem(Handle plugin, int numParams)
...
+		ThrowNativeError(SP_ERROR_NATIVE, "Item id %s is invalid", item);
+
+	DataPack dpCallback = view_as<DataPack>(h_KvItems.GetNum("callbacks", 0));
+	if(dpCallback != null)
+		delete dpCallback;
+	
+	OnItemUnregistered(item_id);
+	
+	h_KvItems.DeleteThis();
+	h_KvItems.Rewind();
```
Comment (TiBarification) [inline addons/sourcemod/scripting/shop/item_manager.sp:599]: I think there are no reason to rewind after we made our actions there.
We trying to keep Rewind each time we are working with KeyValue items.
@R1KO

Comment (IL0co) [reply addons/sourcemod/scripting/shop/item_manager.sp:599]: > I think there are no reason to rewind after we made our actions there.
> We're trying to keep Rewind before each time we are working with KeyValue items.
> @R1KO

during tests, I did a dump (sm_items_dump) of items, without `Rewind()` it was saved somehow crookedly, most likely these are the tricks of `DeleteThis()`...

**Review by TiBarification — APPROVED** [review]

File: addons/sourcemod/scripting/include/shop.inc [hunk 4]
```diff
@@ -127,7 +127,7 @@ public void __pl_shop_SetNTVOptional()
 {
 	MarkNativeAsOptional("Shop_IsStarted");
 	MarkNativeAsOptional("Shop_UnregisterMe");
-	MarkNativeAsOptional("Shop_ShowItemPanel");
```
Comment (R1KO) [inline addons/sourcemod/scripting/include/shop.inc:130]: зачем убрал?

**Review by R1KO — CHANGES_REQUESTED** [review]

File: addons/sourcemod/scripting/include/shop.inc (hunk 4 above)
Comment (IL0co) [reply addons/sourcemod/scripting/include/shop.inc:130]: > зачем убрал?

что-то наклацал и гит это засчитал как изменение...

**Review by R1KO — APPROVED** [review]

**Pull request merged**